src/transformers/trainer.py

```diff
@@ -409,24 +409,34 @@ def __init__(
         # Mixed precision setup
         self.use_apex = False
         self.use_amp = False
-        self.fp16_backend = None
+        self.half_precision_backend = None
 
-        if args.fp16:
-            if args.fp16_backend == "auto":
-                self.fp16_backend = "amp" if _is_native_amp_available else "apex"
+        if args.fp16 or args.bf16:
+            if args.half_precision_backend == "auto":
+                if _is_native_amp_available:
+                    self.half_precision_backend = "amp"
+                else:
+                    if args.bf16:
+                        raise ValueError("Tried to use `bf16` but native amp is not available")
+                    else:
+                        self.half_precision_backend = "apex"
             else:
-                self.fp16_backend = args.fp16_backend
-            logger.info(f"Using {self.fp16_backend} fp16 backend")
-
-        if args.fp16 and not args.deepspeed:  # deepspeed manages its own fp16
-            if self.fp16_backend == "amp":
+                self.half_precision_backend = args.half_precision_backend
+            logger.info(f"Using {self.half_precision_backend} half precision backend")
+        
+        self.do_grad_scaling = False
+        if (args.fp16 or args.bf16) and not args.deepspeed:  # deepspeed manages its own half precision
+            if self.half_precision_backend == "amp":
                 self.use_amp = True
-                if is_sagemaker_mp_enabled():
-                    self.scaler = smp.amp.GradScaler()
-                elif self.sharded_ddp is not None:
-                    self.scaler = ShardedGradScaler()
-                else:
-                    self.scaler = torch.cuda.amp.GradScaler()
+                self.fast_dtype = torch.float16 if args.fp16 else torch.bfloat16
```

Comment:
I think `fast` literally means "fast" in that mixed precision is supposed to increase speed (although it's a bit of an incomplete name in that speed isn't the sole focus of `amp`).  I'm ok with `amp_dtype`, especially because it associates the variable name with `use_amp`.

Reply:
I wonder if it's not too late to propose to remove `fast_` in pytorch, IMHO it's an odd way to name a functional argument.

Reply:
https://github.com/pytorch/pytorch/pull/61002/files#r694163956

Reply:
OK, great, this PR will rename `s/fast_dtype/dtype/` in nightly and also add a utility function to check if bf16 is supported

https://github.com/pytorch/pytorch/pull/63798